### PR TITLE
fix: Using smart pointers instead of raw pointer.

### PR DESCRIPTION
**Pull request by ZipoChan**
Fix: https://github.com/casbin/casbin-cpp/issues/55

Using smart pointers instead of raw pointer,to avoid the leak of memory.
``` Cpp
class Enforcer : public IEnforcer{
    private:

        string model_path;
        shared_ptr<Model> model;
        FunctionMap func_map;
        shared_ptr<Effector> eft;

        shared_ptr<Adapter> adapter;
        shared_ptr<Watcher> watcher;
...
```

### CODE REVIEW

**Comment by hsluoyz** [comment]
@xcaptain @divy9881 please review.

File: casbin/enforcer.cpp [hunk 1]
```diff
@@ -162,8 +162,8 @@ bool Enforcer :: enforce(string matcher, Scope scope) {
 /**
  * Enforcer is the default constructor.
  */
-Enforcer* Enforcer :: NewEnforcer() {
-    Enforcer* e = new Enforcer;
+shared_ptr<Enforcer> Enforcer :: NewEnforcer() {
```
Comment (xcaptain) [inline casbin/enforcer.cpp:165]: How about also change `NewEnforcer` to C++ style constructor? The same question also found in `NewModel` and `NewEffector`

File: casbin/enforcer.cpp [hunk 2]
```diff
@@ -173,8 +173,8 @@ Enforcer* Enforcer :: NewEnforcer() {
  * @param model_path the path of the model file.
  * @param policyFile the path of the policy file.
  */
-Enforcer* Enforcer :: NewEnforcer(string model_path, string policyFile) {
-    return NewEnforcer(model_path, FileAdapter :: NewAdapter(policyFile));
+shared_ptr<Enforcer> Enforcer :: NewEnforcer(string model_path, string policyFile) {
+    return NewEnforcer(model_path, shared_ptr<FileAdapter>(FileAdapter :: NewAdapter(policyFile)));
```
Comment (xcaptain) [inline casbin/enforcer.cpp:177]: Really need to pass in a `shared_ptr<FileAdapter>`? I think the adapter will not be referenced anywhere

File: test/test_enforcer.cpp [hunk 3]
```diff
@@ -40,27 +40,27 @@ namespace test_enforcer
...
+        void TestEnforce(shared_ptr<Enforcer> e, unordered_map<string,string> params, bool res) {
             Assert::AreEqual(res, e->Enforce(params));
         }
 
 
         TEST_METHOD(TestFourParams) {
             string model = filePath("../examples/rbac_with_domains_model.conf");
             string policy = filePath("../examples/rbac_with_domains_policy.csv");
-            Enforcer* e = Enforcer::NewEnforcer(model, policy);
+            shared_ptr<Enforcer> e = Enforcer::NewEnforcer(model, policy);
```
Comment (xcaptain) [inline test/test_enforcer.cpp:63]: How about using `auto` so no need to change so much code when you change a return type?

File: casbin/enforcer.cpp (hunk 1 above)
Comment (divy9881) [reply casbin/enforcer.cpp:165]: Yeah, I agree we must switch to class constructors instead of methods for initialization.

**Comment by divy9881** [comment]
@ZipoChan any progress?

**Comment by hsluoyz** [comment]
@ZipoChan please resolve conflicts.

**Comment by ZipoChan** [comment]
Ok,I will resolve conflicts and use unique_ptr instead of shared_ptr.
@hsluoyz @divy9881

**Comment by hsluoyz** [comment]
@ZipoChan please clear the git history (reset, rebase, cherry-pick, etc.). Don't carry a lot of commit logs from @divy9881 

image

**Comment by ZipoChan** [comment]
@hsluoyz @divy9881 
I have clear the git history.

**Comment by hsluoyz** [comment]
@xcaptain @divy9881

File: casbin/enforcer.cpp [hunk 4]
```diff
@@ -173,8 +173,8 @@ Enforcer* Enforcer :: NewEnforcer() {
  * @param model_path the path of the model file.
  * @param policyFile the path of the policy file.
  */
-Enforcer* Enforcer :: NewEnforcer(string model_path, string policyFile) {
-    return NewEnforcer(model_path, FileAdapter :: NewAdapter(policyFile));
+unique_ptr<Enforcer> Enforcer :: NewEnforcer(string model_path, string policyFile) {
```
Comment (divy9881) [inline casbin/enforcer.cpp:176]: I think indent is not correct over here.

File: test/test_rbac_api.cpp [hunk 5]
```diff
@@ -74,7 +74,7 @@ namespace test_rbac_api
             }
 
             TEST_METHOD(TestEnforcer_AddRolesForUser) {
-                Enforcer* e = Enforcer::NewEnforcer("../../examples/rbac_model.conf", "../../examples/rbac_policy.csv");
+                shared_ptr<Enforcer> e = Enforcer::NewEnforcer("../../examples/rbac_model.conf", "../../examples/rbac_policy.csv");
```
Comment (divy9881) [inline test/test_rbac_api.cpp:77]: I think here, it must be unique_ptr.

File: casbin/enforcer.cpp [hunk 6]
```diff
@@ -162,9 +162,9 @@ bool Enforcer :: enforce(string matcher, Scope scope) {
 /**
  * Enforcer is the default constructor.
  */
-Enforcer* Enforcer :: NewEnforcer() {
-    Enforcer* e = new Enforcer;
-    return e;
+unique_ptr<Enforcer> Enforcer ::NewEnforcer() {
```
Comment (divy9881) [inline casbin/enforcer.cpp:165]: And also @ZipoChan can you replace this class methods to C++ constructors instead.

Comment (divy9881) [reply casbin/enforcer.cpp:165]: I think, this changes can be made in another PR.

File: casbin/enforcer.cpp (hunk 1 above)
Comment (divy9881) [reply casbin/enforcer.cpp]: We can make this changes in any another PR, as it involves making lot of changes.

File: casbin/enforcer.cpp (hunk 6 above)
Comment (xcaptain) [reply casbin/enforcer.cpp:165]: If using a constructor, then don't return a pointer, return the object is ok and use a pointer if want to share in other places.

**Comment by divy9881** [comment]
Use of smart pointer makes accessing the API difficult, of-course there are advantages using this, but makes much more sophisticated for the users,
```
unique_ptr<Enforcer> e = Enforcer :: NewEnforcer()
```
I am not a fan of this.
In the future, we will have to make some aliases to mask this unique_ptr template typing.

**Comment by xcaptain** [comment]
@divy9881 A better way is 

```
auto e = Enforcer();
```

and if this e needs to be used by other functions then pass a smart pointer. What do you think?

**Comment by divy9881** [comment]
Yeah, I agree `auto` is there to serve for dynamic typing, but some C++ projects tend to avoid using auto in their source code itself, I am planning to expose the API like, firstly change class names from `Enforcer, Model, Adapter -> Enforcer_, Model_, Adapter_` and then define aliases of `shared_ptr<Model_> alias Model`, `unique_ptr<Enforcer_> alias Enforcer`

After this, we can instantiate the Enforcer like,
Enforcer e(new Enforcer_());

**Comment by divy9881** [comment]
>@divy9881 A better way is

>auto e = Enforcer();
>and if this e needs to be used by other functions then pass a smart pointer. What do you think?

Can you elaborate, @xcaptain.

**Comment by divy9881** [comment]
I too think, its okay, for now, the discussed changes can be made in other PR.

**Pull request merged**